grudge/dt_utils.py

```diff
@@ -0,0 +1,206 @@
+
+    actx = dcoll._setup_actx
+    volm_discr = dcoll.discr_from_dd(dd)
+
+    if any(not isinstance(grp, SimplexElementGroupBase)
+           for grp in volm_discr.groups):
+        raise NotImplementedError(
+            "Geometric factors are only implemented for simplex element groups"
+        )
+
+    cell_ones = volm_discr.zeros(actx) + 1.0
+    cell_vols = op.elementwise_sum(dcoll, op.mass(dcoll, dd, cell_ones))
+
+    # NOTE: The cell volumes are the *same* at each nodal location.
+    # Take the cell vols at each nodal location and average them to get
+    # a single value per cell.
+    @memoize_in(actx, (dt_geometric_factor, "cell_volume_knl"))
+    def cv_prg():
+        return make_loopy_program(
+            [
+                "{[iel]: 0 <= iel < nelements}",
+                "{[jdof]: 0 <= jdof < n_nodes}"
+            ],
+            """
+                result[iel] = sum(jdof, cell_vol[iel, jdof]) / n_nodes
+            """,
+            name="cell_volume"
+        )
+
+    cell_vols = DOFArray(
+        actx,
+        data=tuple(
+            actx.call_loopy(cv_prg(),
+                            cell_vol=cv_i)["result"]
+
+            for vgrp, cv_i in zip(volm_discr.groups,
+                                  actx.np.fabs(cell_vols))
+        )
+    )
+
+    if dcoll.dim == 1:
+        return op.nodal_min(dcoll, dd, cell_vols)
+
+    dd_face = DOFDesc("all_faces", dd.discretization_tag)
+    face_discr = dcoll.discr_from_dd(dd_face)
+    face_ones = face_discr.zeros(actx) + 1.0
+    face_areas = op.elementwise_sum(
+        dcoll, op._apply_mass_operator(dcoll, dd_face, dd_face, face_ones)
+    )
+
+    # NOTE: The face areas are the *same* at each nodal location.
+    # Take the face areas at each nodal location and average them to get
+    # a single value per face. Then take each face area and compute the
+    # sum over all faces to get the total surface area
+    @memoize_in(actx, (dt_geometric_factor, "total_surface_area_knl"))
+    def sa_prg():
+        return make_loopy_program(
+            [
+                "{[iel]: 0 <= iel < nelements}",
+                "{[f]: 0 <= f < nfaces}",
+                "{[jdof]: 0 <= jdof < nf_nodes}"
+            ],
+            """
+                result[iel] = sum(f, sum(jdof, face_area[f, iel, jdof]) / nf_nodes)
+            """,
+            name="total_surface_area"
+        )
+
+    surface_areas = DOFArray(
+        actx,
+        data=tuple(
+            actx.call_loopy(sa_prg(),
+                            face_area=face_ae_i.reshape(
+                                vgrp.mesh_el_group.nfaces,
+                                vgrp.nelements,
+                                afgrp.nunit_dofs
+                            ))["result"]
+
+            for vgrp, afgrp, face_ae_i in zip(volm_discr.groups,
+                                              face_discr.groups,
+                                              actx.np.fabs(face_areas))
+        )
+    )
```

Comment:
Could be done as an einsum with a post-mutiply.

Reply:
Done: https://github.com/inducer/grudge/pull/108/commits/021fd898e939eebde9dc65e9700e9f9363a59f14